File: src/main/java/org/janelia/saalfeldlab/conversion/ExtractHighestResolutionLabelDataset.java
package org.janelia.saalfeldlab.conversion;

import com.pivovarit.function.ThrowingConsumer;
import gnu.trove.map.TLongLongMap;
import gnu.trove.map.hash.TLongLongHashMap;
import net.imglib2.Cursor;
import net.imglib2.RandomAccessibleInterval;
import net.imglib2.algorithm.util.Grids;
import net.imglib2.converter.Converter;
import net.imglib2.converter.Converters;
import net.imglib2.type.NativeType;
import net.imglib2.type.label.LabelMultisetType;
import net.imglib2.type.numeric.IntegerType;
import net.imglib2.type.numeric.integer.UnsignedLongType;
import net.imglib2.util.Intervals;
import net.imglib2.view.Views;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.janelia.saalfeldlab.label.spark.N5Helpers;
import org.janelia.saalfeldlab.n5.DataType;
import org.janelia.saalfeldlab.n5.DatasetAttributes;
import org.janelia.saalfeldlab.n5.GzipCompression;
import org.janelia.saalfeldlab.n5.N5Reader;
import org.janelia.saalfeldlab.n5.imglib2.N5LabelMultisets;
import org.janelia.saalfeldlab.n5.imglib2.N5Utils;
import org.janelia.saalfeldlab.n5.spark.supplier.N5ReaderSupplier;
import org.janelia.saalfeldlab.n5.spark.supplier.N5WriterSupplier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import picocli.CommandLine;
import scala.Tuple2;

import java.io.IOException;
import java.io.Serializable;
import java.lang.invoke.MethodHandles;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ExtractHighestResolutionLabelDataset  {

    private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

    private static Set<DataType> VALID_TYPES = Stream
            .of(DataType.values())
            .filter(DataType.UINT64::equals)
            .filter(DataType.UINT32::equals)
            .filter(DataType.INT64::equals)
            .collect(Collectors.toSet());

    private static boolean isValidType(final DataType dataType) {
        return VALID_TYPES.contains(dataType);
    }

    private static Set<String> DATASET_ATTRIBUTES = Stream
            .of("dimensions", "blockSize", "compression", "dataType")
            .collect(Collectors.toSet());

    private static class LookupPair implements Serializable {

        public final long key;

        public final long value;

        private LookupPair(long key, long value) {
            this.key = key;
            this.value = value;
        }

        private static class Converter implements CommandLine.ITypeConverter<LookupPair> {

            @Override
            public LookupPair convert(String s) {
                final String[] split = s.split("=");
                return new LookupPair(Long.parseLong(split[0]), Long.parseLong(split[1]));
            }
        }
    }

    public static class Args implements Callable<Void>, Serializable {

        @CommandLine.Option(names = {"--input-container", "-i"}, required=true)
        String inputContainer = null;

        @CommandLine.Option(names = {"--input-dataset", "-I"}, required=true, description = "" +
                "Can be a Paintera dataset, multi-scale N5 group, or regular dataset. " +
                "Highest resolution dataset will be used for Paintera dataset (data/s0) and multi-scale group (s0).")
        String inputDataset = null;

        @CommandLine.Option(names = {"--output-container", "-o"}, required=true)
        String outputContainer = null;

        @CommandLine.Option(names = {"--output-dataset", "-O"}, required=false, description = "defaults to input dataset")
        String outputDataset = null;

        @CommandLine.Option(names = {"--block-size"}, required=false, split = ",", description = "" +
                "Block size for output dataset. Will default to block size of input dataset if not specified.")
        int[] blockSize = null;

        @CommandLine.Option(names = "--consider-fragment-segment-assignment", required = false, defaultValue = "false", description = "" +
                "Consider fragment-segment-assignment inside Paintera dataset. Will be ignored if not a Paintera dataset")
        Boolean considerFragmentSegmentAssignment = false;

        @CommandLine.Option(names = "--additional-assignment", split = ",", required = false, converter = LookupPair.Converter.class, description = "" +
                "Add additional lookup-values in the format `k=v'. Warning: Consistency with fragment-segment-assignment is not ensured.")
        LookupPair[] additionalAssignments = null;

        @Override
        public Void call() throws IOException {
            outputDataset = outputDataset == null ? inputDataset : outputDataset;

            if (inputContainer.equals(outputContainer) && inputDataset.equals(outputDataset)) {
                throw new IOException(String.format(
                        "Output dataset %s would overwrite input dataset %s in output container %s (same as input container %s)",
                        outputDataset,
                        inputDataset,
                        outputContainer,
                        inputContainer));
            }

            final SparkConf conf = new SparkConf().setAppName(MethodHandles.lookup().lookupClass().getName());

            final TLongLongMap assignment = new TLongLongHashMap();
            if (additionalAssignments != null)
                for (final LookupPair pair: additionalAssignments)
                    assignment.put(pair.key, pair.value);

            try (final JavaSparkContext sc = new JavaSparkContext(conf)) {
                extract(
                        sc,
                        () -> N5Helpers.n5Reader(inputContainer),
                        () -> N5Helpers.n5Writer(outputContainer),
                        inputDataset,
                        outputDataset,
                        blockSize,
                        considerFragmentSegmentAssignment != null && considerFragmentSegmentAssignment,
                        assignment);
            }

            return null;
        }
    }

    public static void main(String[] args) {
        CommandLine.call(new Args(), args);
    }

    public static <IN extends NativeType<IN> & IntegerType<IN>> void extract(
            final JavaSparkContext sc,
            final N5ReaderSupplier n5in,
            final N5WriterSupplier n5out,
            final String datasetIn,
            final String datasetOut,
            final int[] blockSizeOut,
            final boolean considerFragmentSegmentAssignment,
            final TLongLongMap assignment) throws IOException {
        ExtractHighestResolutionLabelDataset.<IN, UnsignedLongType>extract(
                sc,
                n5in,
                n5out,
                datasetIn,
                datasetOut,
                blockSizeOut,
                (Serializable & Supplier< UnsignedLongType >) UnsignedLongType::new,
                Collections.emptyMap(),
                considerFragmentSegmentAssignment,
                assignment);
    }

    public static <IN extends NativeType<IN> & IntegerType<IN>, OUT extends NativeType<OUT> & IntegerType<OUT>> void extract(
            final JavaSparkContext sc,
            final N5ReaderSupplier n5in,
            final N5WriterSupplier n5out,
            final String datasetIn,
            final String datasetOut,
            final int[] blockSizeOut,
            final Supplier<OUT> outputTypeSupplier,
            final Map<String, Object> additionalAttributes,
            final boolean considerFragmentSegmentAssignment,
            final TLongLongMap assignment
            ) throws IOException {

        if (!n5in.get().exists(datasetIn)) {
            throw new IOException(String.format("%s does not exist in container %s", datasetIn, n5in.get()));
        }

        if (!n5in.get().datasetExists(datasetIn)) {
            if (n5in.get().listAttributes(datasetIn).containsKey("painteraData")) {
                try {
                    final Map<String, Object> updatedAdditionalEntries = new HashMap<>(additionalAttributes);
                    Optional.ofNullable(n5in.get().getAttribute(datasetIn, "maxId", long.class)).ifPresent(id -> updatedAdditionalEntries.put("maxId", id));
                    if (considerFragmentSegmentAssignment) {
                        final TLongLongMap loadedAssignments = readAssignments(n5in.get(), datasetIn + "/fragment-segment-assignment");
                        loadedAssignments.putAll(assignment);
                        assignment.clear();
                        assignment.putAll(loadedAssignments);
                    }
                    extract(sc, n5in, n5out, datasetIn + "/data", datasetOut, blockSizeOut, outputTypeSupplier, updatedAdditionalEntries, considerFragmentSegmentAssignment, assignment);
                    return;
                } catch (final NoValidDatasetException e) {
                    throw new NoValidDatasetException(n5in.get(), datasetIn);
                }
            }
            else if (n5in.get().exists(datasetIn + "/s0")) {
                try {
                    extract(sc, n5in, n5out, datasetIn + "/s0", datasetOut, blockSizeOut, outputTypeSupplier, additionalAttributes, considerFragmentSegmentAssignment, assignment);
                    return;
                } catch (final NoValidDatasetException e) {
                    throw new NoValidDatasetException(n5in.get(), datasetIn);
                }
            }
            else
                throw new NoValidDatasetException(n5in.get(), datasetIn);
        }

        final boolean outputIsLabelMultiset = outputTypeSupplier.get() instanceof LabelMultisetType;

        final DatasetAttributes attributesIn = n5in.get().getDatasetAttributes(datasetIn);
        final long[] dimensions = attributesIn.getDimensions().clone();
        final int[] blockSize = blockSizeOut == null ? attributesIn.getBlockSize() : blockSizeOut;
        final DataType dataType = outputIsLabelMultiset
                ? DataType.UINT8
                : N5Utils.dataType(outputTypeSupplier.get());

        n5out.get().createDataset(datasetOut, dimensions, blockSize, dataType, new GzipCompression());
        final long[] keys = assignment.keys();
        final long[] values = assignment.values();

        // TODO automate copy of attributes if/when N5 separates attributes from dataset attributes
//        for (Map.Entry<String, Class<?>> entry :n5in.get().listAttributes(datasetIn).entrySet()) {
//            if (DATASET_ATTRIBUTES.contains(entry.getKey()))
//                continue;
//            try {
//                Object attr = n5in.get().getAttribute(datasetIn, entry.getKey(), entry.getValue());
//                LOG.debug("Copying attribute { {}: {} } of type {}", entry.getKey(), attr, entry.getValue());
//                n5out.get().setAttribute(datasetOut, entry.getKey(), attr);
//            } catch (IOException e) {
//                LOG.warn("Unable to copy attribute { {}: {} }", entry.getKey(), entry.getValue());
//                LOG.debug("Unable to copy attribute { {}: {} }", entry.getKey(), entry.getValue(), e);
//            }
//        }

        Optional
                .ofNullable(n5in.get().getAttribute(datasetIn, "resolution", double[].class))
                .ifPresent(ThrowingConsumer.unchecked(r -> n5out.get().setAttribute(datasetOut, "resolution", r)));

        Optional
                .ofNullable(n5in.get().getAttribute(datasetIn, "offset", double[].class))
                .ifPresent(ThrowingConsumer.unchecked(o -> n5out.get().setAttribute(datasetOut, "offset", o)));

        Optional
                .ofNullable(n5in.get().getAttribute(datasetIn, "maxId", long.class))
                .ifPresent(ThrowingConsumer.unchecked(id -> n5out.get().setAttribute(datasetOut, "maxId", id)));

        additionalAttributes.entrySet().forEach(ThrowingConsumer.unchecked(e -> n5out.get().setAttribute(datasetOut, e.getKey(), e.getValue())));

        try {
            n5out.get().setAttribute(datasetOut, N5LabelMultisets.LABEL_MULTISETTYPE_KEY, outputIsLabelMultiset);
        } catch (IOException e) {
            LOG.warn("Unable to write attribute { {}: {} }", N5LabelMultisets.LABEL_MULTISETTYPE_KEY, outputIsLabelMultiset);
            LOG.debug("Unable to write attribute { {}: {} }", N5LabelMultisets.LABEL_MULTISETTYPE_KEY, outputIsLabelMultiset, e);
        }
        final boolean isLabelMultiset = N5LabelMultisets.isLabelMultisetType(n5in.get(), datasetIn);

        if (!(DataType.UINT8.equals(attributesIn.getDataType()) && isLabelMultiset || isValidType(attributesIn.getDataType()) && !isLabelMultiset))
            throw new InvalidTypeException(attributesIn.getDataType(), isLabelMultiset);

        final List<Tuple2<Tuple2<long[], long[]>, long[]>> blocks = Grids
                .collectAllContainedIntervalsWithGridPositions(dimensions, blockSize)
                .stream()
                .map(p -> new Tuple2<>(new Tuple2<>(Intervals.minAsLongArray(p.getA()), Intervals.maxAsLongArray(p.getA())), p.getB()))
                .collect(Collectors.toList());

        sc
                .parallelize(blocks)
                .foreach(blockWithPosition -> {
                    RandomAccessibleInterval<IN> input = isLabelMultiset
                            ? (RandomAccessibleInterval) N5LabelMultisets.openLabelMultiset(n5in.get(), datasetIn)
                            : N5Utils.open(n5in.get(), datasetIn);
                    final RandomAccessibleInterval<IN> block = Views.interval(
                            input,
                            blockWithPosition._1()._1(),
                            blockWithPosition._1()._2());

                    final DatasetAttributes attributes = new DatasetAttributes(
                            dimensions,
                            blockSize,
                            N5Utils.dataType(outputTypeSupplier.get()),
                            new GzipCompression());

                    final RandomAccessibleInterval<OUT> converted = Converters.convert(block, getAppropriateConverter(new TLongLongHashMap(keys, values)), outputTypeSupplier.get());

                    N5Utils.saveBlock(converted, n5out.get(), datasetOut, attributes, blockWithPosition._2());

                });

    }

    private static <IN extends IntegerType<IN>, OUT extends IntegerType<OUT>> Converter<IN, OUT> getAppropriateConverter(final TLongLongMap map) {
        LOG.trace("Getting converter for map {}", map);
        if (map == null || map.isEmpty())
            return (s, t) -> t.setInteger(s.getIntegerLong());
        return (s, t) -> {
            final long k = s.getIntegerLong();
            if (map.containsKey(k))
                t.setInteger(map.get(k));
            else
                t.setInteger(k);
        };
    }

    private static class NoValidDatasetException extends IOException {

        private NoValidDatasetException(final N5Reader container, final String dataset) {
            super(String.format("Unable to find valid data at %s in container %s", dataset, container));
        }

        private NoValidDatasetException(final String message) {
            super(message);
        }

    }

    private static class InvalidTypeException extends NoValidDatasetException {

        private InvalidTypeException(final DataType dataType, final boolean isLabelMultiset) {

            super(String.format(
                    "Not a valid data type for conversion: (DataType=%s, isLabelMultiset=%s). Expected (DataType=%s, isLabelMultiset=true) or (DataType=any from %s, isLabelMultiset=false)",
                    dataType,
                    isLabelMultiset,
                    DataType.UINT8,
                    VALID_TYPES));

        }

    }

    private static TLongLongMap readAssignments(
            final N5Reader container,
            final String dataset) {
        try {
            RandomAccessibleInterval<UnsignedLongType> data = openDatasetSafe(container, dataset);
            final long[] keys = new long[(int) data.dimension(0)];
            final long[] values = new long[keys.length];
            LOG.debug("Found {} assignments", keys.length);
            final Cursor<UnsignedLongType> keyCursor = Views.flatIterable(Views.hyperSlice(data, 1, 0L)).cursor();
            final Cursor<UnsignedLongType> valueCursor = Views.flatIterable(Views.hyperSlice(data, 1, 1L)).cursor();
            for (int i = 0; i < keys.length; ++i) {
                keys[i] = keyCursor.next().getIntegerLong();
                values[i] = valueCursor.next().getIntegerLong();
            }
            return new TLongLongHashMap(keys, values);
        } catch (IOException e) {
            LOG.debug("Exception while trying to return initial lut from N5", e);
            LOG.info("Unable to read initial lut from {} in {} -- returning empty map", dataset, container);
            return new TLongLongHashMap();
        }
    }

    private static RandomAccessibleInterval<UnsignedLongType> openDatasetSafe(
            final N5Reader reader,
            final String dataset
    ) throws IOException {
        return DataType.UINT64.equals(reader.getDatasetAttributes(dataset).getDataType())
                ? N5Utils.open(reader, dataset)
                : openAnyIntegerTypeAsUnsignedLongType(reader, dataset);
    }

    private static <T extends IntegerType<T> & NativeType<T>> RandomAccessibleInterval<UnsignedLongType> openAnyIntegerTypeAsUnsignedLongType(
            final N5Reader reader,
            final String dataset
    ) throws IOException {
        final RandomAccessibleInterval<T> img = N5Utils.open(reader, dataset);
        return Converters.convert(img, (s, t) -> t.setInteger(s.getIntegerLong()), new UnsignedLongType());
    }

}

File: src/main/kotlin/org/janelia/saalfeldlab/conversion/ExtractHighestResolutionLabelDataset.kt
package org.janelia.saalfeldlab.conversion

import com.pivovarit.function.ThrowingConsumer
import gnu.trove.map.TLongLongMap
import gnu.trove.map.hash.TLongLongHashMap
import io.github.oshai.kotlinlogging.KotlinLogging
import net.imglib2.Interval
import net.imglib2.RandomAccessibleInterval
import net.imglib2.algorithm.util.Grids
import net.imglib2.img.array.ArrayImgFactory
import net.imglib2.algorithm.util.Singleton
import net.imglib2.algorithm.util.Singleton.ThrowingSupplier
import net.imglib2.converter.Converter
import net.imglib2.converter.Converters
import net.imglib2.type.NativeType
import net.imglib2.type.label.LabelMultisetType
import net.imglib2.type.numeric.IntegerType
import net.imglib2.type.numeric.integer.UnsignedLongType
import net.imglib2.util.Intervals
import net.imglib2.util.Pair
import net.imglib2.view.Views
import org.apache.http.client.utils.URIBuilder
import org.apache.http.message.BasicNameValuePair
import org.apache.spark.api.java.JavaSparkContext
import org.janelia.saalfeldlab.n5.DataType
import org.janelia.saalfeldlab.n5.DatasetAttributes
import org.janelia.saalfeldlab.n5.N5Reader
import org.janelia.saalfeldlab.n5.imglib2.N5LabelMultisets
import org.janelia.saalfeldlab.n5.imglib2.N5Utils
import org.janelia.saalfeldlab.n5.universe.metadata.axes.Axis
import org.janelia.saalfeldlab.n5.universe.metadata.ome.ngff.OmeNgffMetadata
import org.janelia.saalfeldlab.n5.universe.metadata.ome.ngff.OmeNgffMetadataParser
import org.janelia.saalfeldlab.n5.zarr.ZarrKeyValueWriter
import org.janelia.saalfeldlab.n5.zarr.v3.ZarrV3DatasetAttributes
import org.janelia.saalfeldlab.n5.zarr.v3.ZarrV3KeyValueWriter
import org.janelia.saalfeldlab.n5.spark.downsample.N5LabelDownsamplerSpark
import org.janelia.saalfeldlab.n5.spark.supplier.N5ReaderSupplier
import org.janelia.saalfeldlab.n5.spark.supplier.N5WriterSupplier
import org.janelia.scicomp.n5.zstandard.ZstandardCompression
import picocli.CommandLine
import scala.Tuple2
import java.io.IOException
import java.io.Serializable
import java.nio.ByteBuffer
import java.util.Optional
import java.util.function.Supplier
import java.util.stream.Collectors

object ExtractHighestResolutionLabelDataset {
	private val LOG = KotlinLogging.logger { }

	private val VALID_TYPES: Set<DataType> = setOf(
		DataType.UINT64, DataType.UINT32, DataType.INT64
	)

	private fun isValidType(dataType: DataType): Boolean {
		return VALID_TYPES.contains(dataType)
	}

	@Throws(IOException::class)
	fun extractNoGenerics(
		sc: JavaSparkContext,
		n5in: N5ReaderSupplier,
		n5out: N5WriterSupplier,
		datasetIn: String?,
		datasetOut: String?,
		blockSizeOut: IntArray?,
		considerFragmentSegmentAssignment: Boolean,
		assignment: TLongLongMap,
		xyzUnit: Array<String>,
		chunksPerShard: IntArray? = null,
		scales: Array<IntArray> = emptyArray(),
		downsampleBlockSizes: Array<IntArray> = emptyArray(),
		resolutionOverride: DoubleArray? = null,
		offsetOverride: DoubleArray? = null
	) {
		extract(sc, n5in, n5out, datasetIn, datasetOut, blockSizeOut, considerFragmentSegmentAssignment, assignment, xyzUnit, chunksPerShard, scales, downsampleBlockSizes, resolutionOverride, offsetOverride)
	}

	@JvmStatic
	@Throws(IOException::class)
	fun <IN> extract(
		sc: JavaSparkContext,
		n5in: N5ReaderSupplier,
		n5out: N5WriterSupplier,
		datasetIn: String?,
		datasetOut: String?,
		blockSizeOut: IntArray?,
		considerFragmentSegmentAssignment: Boolean,
		assignment: TLongLongMap,
		xyzUnit: Array<String>,
		chunksPerShard: IntArray? = null,
		scales: Array<IntArray> = emptyArray(),
		downsampleBlockSizes: Array<IntArray> = emptyArray(),
		resolutionOverride: DoubleArray? = null,
		offsetOverride: DoubleArray? = null
	) where IN : NativeType<IN>?, IN : IntegerType<IN>? {
		extract<IN, UnsignedLongType>(
			sc,
			n5in,
			n5out,
			datasetIn,
			datasetOut,
			blockSizeOut,
			object : Supplier<UnsignedLongType>, Serializable {
				override fun get() = UnsignedLongType()
			},
			emptyMap(),
			considerFragmentSegmentAssignment,
			assignment,
			xyzUnit,
			chunksPerShard,
			scales,
			downsampleBlockSizes,
			resolutionOverride,
			offsetOverride
		)
	}

	@Throws(IOException::class)
	fun <IN, OUT> extract(
		sc: JavaSparkContext,
		n5in: N5ReaderSupplier,
		n5out: N5WriterSupplier,
		datasetIn: String?,
		datasetOut: String?,
		blockSizeOut: IntArray?,
		outputTypeSupplier: Supplier<OUT>,
		additionalAttributes: Map<String?, Any>,
		considerFragmentSegmentAssignment: Boolean,
		assignment: TLongLongMap,
		xyzUnit: Array<String>,
		chunksPerShard: IntArray? = null,
		scales: Array<IntArray> = emptyArray(),
		downsampleBlockSizes: Array<IntArray> = emptyArray(),
		resolutionOverride: DoubleArray? = null,
		offsetOverride: DoubleArray? = null
	) where IN : NativeType<IN>?, IN : IntegerType<IN>?, OUT : NativeType<OUT>?, OUT : IntegerType<OUT>? {
		val n5InLocal = n5in.get()
		if (!n5InLocal.exists(datasetIn)) {
			throw IOException(String.format("%s does not exist in container %s", datasetIn, n5InLocal))
		}

		if (!n5InLocal.datasetExists(datasetIn)) {
			if (n5InLocal.listAttributes(datasetIn).containsKey("painteraData")) {
				try {
					val updatedAdditionalEntries: MutableMap<String?, Any> = HashMap(additionalAttributes)
					Optional.ofNullable(n5InLocal.getAttribute(datasetIn, "maxId", Long::class.javaPrimitiveType)).ifPresent { id: Long -> updatedAdditionalEntries["maxId"] = id }
					if (considerFragmentSegmentAssignment) {
						val loadedAssignments = readAssignments(n5InLocal, "$datasetIn/fragment-segment-assignment")
						loadedAssignments.putAll(assignment)
						assignment.clear()
						assignment.putAll(loadedAssignments)
					}
					extract(
						sc, n5in, n5out, "$datasetIn/data", datasetOut, blockSizeOut, outputTypeSupplier, updatedAdditionalEntries,
						considerFragmentSegmentAssignment, assignment, xyzUnit, chunksPerShard, scales, downsampleBlockSizes, resolutionOverride, offsetOverride
					)
					return
				} catch (e: NoValidDatasetException) {
					throw NoValidDatasetException(n5InLocal, datasetIn)
				}
			} else if (n5InLocal.exists("$datasetIn/s0")) {
				try {
					extract(
						sc, n5in, n5out, "$datasetIn/s0", datasetOut, blockSizeOut, outputTypeSupplier, additionalAttributes,
						considerFragmentSegmentAssignment, assignment, xyzUnit, chunksPerShard, scales, downsampleBlockSizes, resolutionOverride, offsetOverride
					)
					return
				} catch (e: NoValidDatasetException) {
					throw NoValidDatasetException(n5InLocal, datasetIn)
				}
			} else throw NoValidDatasetException(n5InLocal, datasetIn)
		}

		val outputIsLabelMultiset = outputTypeSupplier.get() is LabelMultisetType

		val attributesIn = n5InLocal.getDatasetAttributes(datasetIn)
		val dimensions = attributesIn.dimensions.clone()
		val blockSize = blockSizeOut ?: attributesIn.blockSize
		val dataType = DataType.UINT8.takeIf { outputIsLabelMultiset } ?: N5Utils.dataType(outputTypeSupplier.get())

		val outWriter = n5out.get()
		/* serialize OME-NGFF multiscale metadata; 0.5 unless zarr2 */
		val ngffVersion = "0.5".takeUnless {  outWriter is ZarrKeyValueWriter } ?: "0.4"
		val dataDataset = "$datasetOut/s0"
		outWriter.createGroup(datasetOut)

		val datasetAttributes = if (chunksPerShard != null && outWriter is ZarrV3KeyValueWriter) {
			/* shard size = chunks-per-shard * block size ( when sharded the block size parameter is used for chunk size; maybe should rename)  */
			val shardSize = IntArray(blockSize.size) { chunksPerShard[it] * blockSize[it] }
			ZarrV3DatasetAttributes(dimensions, shardSize, blockSize, dataType, ZstandardCompression())
		} else {
			DatasetAttributes(dimensions, blockSize, dataType, ZstandardCompression())
		}

		outWriter.createDataset( dataDataset, datasetAttributes )
		/* the unit of parallel write is always the DatasetAttributes#blockSize. when sharded this is the shard size */
		val outputBlockSize = outWriter.getDatasetAttributes(dataDataset).blockSize
		val keys = assignment.keys()
		val values = assignment.values()

		val resolution = resolutionOverride ?: n5InLocal.getAttribute(datasetIn, "resolution", DoubleArray::class.java) ?: DoubleArray(dimensions.size) { 1.0 }
		val offset = offsetOverride ?: n5InLocal.getAttribute(datasetIn, "offset", DoubleArray::class.java) ?: DoubleArray(dimensions.size) { 0.0 }
		/* resolution/offset become the OME-NGFF scale/translation transforms; one dataset per pyramid level, with the
		 * per-level scale accumulating the relative downsampling factors */
		val axes = arrayOf(
			Axis(Axis.SPACE, "x", xyzUnit[0], false),
			Axis(Axis.SPACE, "y", xyzUnit[1], false),
			Axis(Axis.SPACE, "z", xyzUnit[2], false)
		)
		val scalePaths = Array(scales.size + 1) { "s$it" }
		val levelResolutions = arrayListOf(resolution.copyOf())
		val levelTranslations = arrayListOf(offset.copyOf())
		val cumulativeFactor = DoubleArray(dimensions.size) { 1.0 }
		scales.forEach { factor ->
			for (d in cumulativeFactor.indices) cumulativeFactor[d] = cumulativeFactor[d] * factor[d]
			levelResolutions.add(DoubleArray(dimensions.size) { resolution[it] * cumulativeFactor[it] })
			levelTranslations.add(offset.copyOf())
		}
		val metadata = OmeNgffMetadata.buildForWriting(
			dimensions.size,
			datasetOut,
			ngffVersion,
			axes,
			scalePaths,
			levelResolutions.toTypedArray(),
			levelTranslations.toTypedArray()
		)

		n5InLocal.getAttribute(datasetIn, "maxId", Long::class.javaPrimitiveType)?.let { maxId ->
			outWriter.setAttribute(dataDataset, "maxId", maxId)
		}

		additionalAttributes.entries.forEach(ThrowingConsumer.unchecked { e: Map.Entry<String?, Any> -> outWriter.setAttribute(dataDataset, e.key, e.value) })

		try {
			outWriter.setAttribute(dataDataset, N5LabelMultisets.LABEL_MULTISETTYPE_KEY, outputIsLabelMultiset)
		} catch (e: IOException) {
			LOG.warn { "Unable to write attribute { ${N5LabelMultisets.LABEL_MULTISETTYPE_KEY}: $outputIsLabelMultiset }" }
			LOG.debug(e) { "Unable to write attribute { ${N5LabelMultisets.LABEL_MULTISETTYPE_KEY}: $outputIsLabelMultiset }" }
		}
		val isLabelMultiset = N5LabelMultisets.isLabelMultisetType(n5InLocal, datasetIn)

		if (!(DataType.UINT8 == attributesIn.dataType && isLabelMultiset || isValidType(attributesIn.dataType) && !isLabelMultiset))
			throw InvalidTypeException(attributesIn.dataType, isLabelMultiset)

		val blocks: List<Tuple2<Tuple2<LongArray, LongArray>, LongArray>> = Grids
			.collectAllContainedIntervalsWithGridPositions(dimensions, outputBlockSize)
			.stream()
			.map { p: Pair<Interval, LongArray> -> Tuple2(Tuple2(Intervals.minAsLongArray(p.a), Intervals.maxAsLongArray(p.a)), p.b) }
			.collect(Collectors.toList())

		sc
			.parallelize(blocks)
			.foreach { blockWithPosition: Tuple2<Tuple2<LongArray, LongArray>, LongArray> ->

				val n5Local = n5in.get()
				val imgCacheKey = URIBuilder(n5Local.uri).setParameters(
					BasicNameValuePair("call", "extract-highest-resolution-label-dataset"),
					BasicNameValuePair("dataset", datasetIn)
				).toString()

				val input: RandomAccessibleInterval<IN> = Singleton.get(imgCacheKey, ThrowingSupplier {
					if (isLabelMultiset) N5LabelMultisets.openLabelMultiset(n5Local, datasetIn) as RandomAccessibleInterval<IN>
					else N5Utils.open(n5Local, datasetIn)
				})

				val block: RandomAccessibleInterval<IN> = Views.interval(
					input,
					blockWithPosition._1()._1(),
					blockWithPosition._1()._2()
				)

				val converted = Converters.convert(
					block,
					getAppropriateConverter(TLongLongHashMap(keys, values)),
					outputTypeSupplier.get()
				)

				val n5LocalOut = n5out.get()
				val writerCacheKey = URIBuilder(n5LocalOut.uri).setParameters(
					BasicNameValuePair("call", "extract-highest-resolution-label-dataset"),
					BasicNameValuePair("type", "writer")
				).toString()

				val writer = Singleton.get(writerCacheKey, ThrowingSupplier { n5LocalOut })

				/* read the created dataset's attributes so sharded writes route into shards */
				val attributes = writer.getDatasetAttributes(dataDataset)

				val fillValue = (attributes as? ZarrV3DatasetAttributes)?.run { ByteBuffer.wrap(fillBytes).long } ?: 0L
				val fillType = outputTypeSupplier.get()!!.also { it.setInteger(fillValue) }
				/* Converters.convert view over LabelMultisetType is not PrimitiveBlocks-compatible,
				 * since LabelMultisetType only pretends to be a NativeType. We have to materialize
				 * it here before saving  */
				val materialized = ArrayImgFactory(fillType).create(*Intervals.dimensionsAsLongArray(converted))
				val source = Views.flatIterable(converted).cursor()
				val target = Views.flatIterable(materialized).cursor()

				while (target.hasNext())
					target.next().setInteger(source.next().integerLong)

				/* saveNonEmptyBlock skips empty chunks/shard. Its gridOffset is in chunk-grid units,
				 * but our parallel unit is the shard, so scale the grid position by chunks-per-shard. */
				val chunkGridOffset = LongArray(dimensions.size) {
					blockWithPosition._2()[it] * (attributes.blockSize[it] / attributes.chunkSize[it])
				}
				N5Utils.saveNonEmptyBlock(materialized, writer, dataDataset, attributes, chunkGridOffset, fillType)
			}

		/* build the downsampled pyramid s1..sN from s0; each level is sharded when chunksPerShard is set */
		var previousLevel = dataDataset
		scales.forEachIndexed { level, factor ->
			val nextLevel = "$datasetOut/s${level + 1}"
			N5LabelDownsamplerSpark.downsampleLabel<UnsignedLongType>(
				sc, n5out, previousLevel, nextLevel, factor,
				downsampleBlockSizes.getOrElse(level) { blockSize }, chunksPerShard, false
			)
			previousLevel = nextLevel
		}

		/* write the multiscale metadata once every level exists */
		OmeNgffMetadataParser(outWriter).writeMetadata(metadata, outWriter, datasetOut)
	}

	private fun <IN : IntegerType<IN>?, OUT : IntegerType<OUT>?> getAppropriateConverter(map: TLongLongMap?): Converter<IN, OUT> {
		LOG.trace { "Getting converter for map $map" }
		if (map == null || map.isEmpty) return Converter { s: IN, t: OUT -> t!!.setInteger(s!!.integerLong) }
		return Converter { s: IN, t: OUT ->
			val k = s!!.integerLong
			if (map.containsKey(k)) t!!.setInteger(map[k])
			else t!!.setInteger(k)
		}
	}

	private fun readAssignments(
		container: N5Reader,
		dataset: String
	): TLongLongMap {
		try {
			val data = openDatasetSafe(container, dataset)
			val keys = LongArray(data.dimension(0).toInt())
			val values = LongArray(keys.size)
			LOG.debug { "Found ${keys.size} assignments" }
			val keyCursor = Views.flatIterable(Views.hyperSlice(data, 1, 0L)).cursor()
			val valueCursor = Views.flatIterable(Views.hyperSlice(data, 1, 1L)).cursor()
			for (i in keys.indices) {
				keys[i] = keyCursor.next().integerLong
				values[i] = valueCursor.next().integerLong
			}
			return TLongLongHashMap(keys, values)
		} catch (e: IOException) {
			LOG.debug(e) { "Exception while trying to return initial lut from N5" }
			LOG.info { "Unable to read initial lut from $dataset in $container -- returning empty map" }
			return TLongLongHashMap()
		}
	}

	@Throws(IOException::class)
	private fun openDatasetSafe(
		reader: N5Reader,
		dataset: String
	): RandomAccessibleInterval<UnsignedLongType> {
		return if (DataType.UINT64 == reader.getDatasetAttributes(dataset).dataType) N5Utils.open(reader, dataset)
		else openAnyIntegerTypeAsUnsignedLongType(reader, dataset)
	}

	@Throws(IOException::class)
	private fun <T> openAnyIntegerTypeAsUnsignedLongType(
		reader: N5Reader,
		dataset: String
	): RandomAccessibleInterval<UnsignedLongType> where T : IntegerType<T>?, T : NativeType<T>? {
		val img: RandomAccessibleInterval<T> = N5Utils.open(reader, dataset)
		return Converters.convert(img, { s: T, t: UnsignedLongType -> t.setInteger(s!!.integerLong) }, UnsignedLongType())
	}

	class LookupPair private constructor(val key: Long, val value: Long) : Serializable {
		class Converter : CommandLine.ITypeConverter<LookupPair> {
			override fun convert(s: String): LookupPair {
				val split = s.split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
				return LookupPair(split[0].toLong(), split[1].toLong())
			}
		}
	}

	private open class NoValidDatasetException : IOException {
		constructor(container: N5Reader, dataset: String?) : super(String.format("Unable to find valid data at %s in container %s", dataset, container))

		protected constructor(message: String) : super(message)
	}

	private class InvalidTypeException(dataType: DataType, isLabelMultiset: Boolean) : NoValidDatasetException(
		String.format(
			"Not a valid data type for conversion: (DataType=%s, isLabelMultiset=%s). Expected (DataType=%s, isLabelMultiset=true) or (DataType=any from %s, isLabelMultiset=false)",
			dataType,
			isLabelMultiset,
			DataType.UINT8,
			VALID_TYPES
		)
	)
}
